lib/List.dbl

```diff
@@ -1,33 +1,158 @@
+  match xs with 
+  | []      => None
+  | x :: xs => Some x
   end
 
-pub let rec filter f xs =
-  match xs with
-  | []      => []
-  | x :: xs => if f x then x :: filter f xs else filter f xs
+pub let tl_opt xs =
+  match xs with 
+  | []      => None
+  | x :: xs => Some xs
   end
 
+pub let nth {`re : {type X} -> Unit ->[|_] X} xs (n : Int) = 
+  let rec nth_aux xs (n : Int) =
+    match xs with 
+    | []      => `re () 
+    | x :: xs => if n == 0 then x else nth_aux xs (n - 1)
+    end 
+  in if n < 0 then `re () else nth_aux xs 0
+  
+pub let nth_opt xs (n : Int) = 
+  let rec nth_opt_aux xs (n : Int) =
+    match xs with 
+    | []      => None 
+    | x :: xs => if n == 0 then Some x else nth_opt_aux xs (n - 1)
+    end 
+  in if n < 0 then None else nth_opt_aux xs n
+
 pub let rec append xs ys =
   match xs with
   | []      => ys
   | x :: xs => x :: append xs ys
   end
 
+pub let rec rev_append xs acc = 
+  match xs with 
+  | []      => acc 
+  | x :: xs => rev_append xs (x :: acc)
+  end 
+
+pub let rev xs = rev_append xs []
+
 pub let rec concat xss =
   match xss with
   | []        => []
   | xs :: xss => append xs (concat xss)
   end
 
+pub let flatten = concat
+
+pub let rec map f xs =
+  match xs with
+  | []      => []
+  | x :: xs => f x :: map f xs
+  end
+
+pub let mapi f xs = 
+  let rec mapi_aux (i : Int) f xs =
+    match xs with 
+    | []      => []
+    | x :: xs => (f i x) :: mapi_aux (i+1) f xs
+    end 
+  in mapi_aux 0 f xs
+
+pub let rec rev_map_append f xs ys =
+  match xs with 
+  | []      => ys 
+  | x :: xs => rev_map_append f xs (f x :: ys)
+  end 
+
+pub let rev_map f xs = rev_map_append f xs []
+
+pub let rec filter f xs =
+  match xs with
+  | []      => []
+  | x :: xs => if f x then x :: filter f xs else filter f xs
+  end
+
+pub let filteri p xs = 
+  let rec filteri_aux p (i : Int) xs =
+    match xs with 
+    | []      => [] 
+    | x :: xs => if p i x then x :: filteri_aux p (i + 1) xs else filteri_aux p (i + 1) xs
+    end 
+  in filteri_aux p 0 xs
+
+pub let rec filter_map f xs =
+  match xs with 
+  | []      => []
+  | x :: xs => 
+    match f x with 
+    | None   => filter_map f xs 
+    | Some x => x :: filter_map f xs 
+    end 
+  end
+
+pub let concat_map f xs = concat (map f xs)
+
+pub let rec take (n : Int) xs = 
+  match xs with 
+  | []      => [] 
+  | x :: xs => if n == 0 then [] else x :: take (n - 1) xs 
+  end
+
+pub let rec drop (n : Int) xs =
+  if n <= 0 then xs else 
+    match xs with 
+    | []      => [] 
+    | x :: xs => drop (n - 1) xs 
+    end 
+
+pub let rec take_while p xs =
+  match xs with 
+  | []      => []
+  | x :: xs => if p x then x :: take_while p xs else []
+  end 
+
+pub let rec drop_while p xs =
+  match xs with 
+  | []       => [] 
+  | x :: xs' => if p x then drop_while p xs' else xs 
+  end 
+
 pub let rec iter f xs =
   match xs with
   | []      => ()
   | x :: xs => let () = f x in iter f xs
   end
 
+pub let rec iteri (i : Int) f xs =
```

Comment:
Should `iteri` take two instead of three parameters? Of course, its implementation needs this additional accumulator, but should it be part of functions signature? Optional parameters (#28) would help here, but they are not implemented yet.